L10-c++-reactive-streams.md

```diff
@@ -0,0 +1,387 @@
+
+
+#### Server streaming: FizzBuzz
+
+```proto
+service TestService {
+  rpc FizzBuzz (FizzBuzzRequest) returns (stream FizzBuzzResponse) {}
+}
+
+message FizzBuzzRequest {
+  int32 upto = 1;
+}
+
+message FizzBuzzResponse {
+  string data = 1;
+}
+```
+
+```cpp
+class TestServiceImpl : public TestService::ReactiveService {
+ public:
+  AnyPublisher<FizzBuzzResponse> FizzBuzz(FizzBuzzRequest request) override {
+    return AnyPublisher<FizzBuzzResponse>(Pipe(
+        Range(1, request.upto()),
+        Map(&FizzBuzz),
+        Map(&MakeFizzBuzzResponse));
+  }
+
+ private:
+  static std::string FizzBuzz(int num) {
+    if ((num % 3) == 0 && (num % 5) == 0) {
+      return "FizzBuzz";
+    } else if ((num % 3) == 0) {
+      return "Fizz";
+    } else if ((num % 5) == 0) {
+      return "Buzz";
+    } else {
+      return std::to_string(num);
+    }
+  }
+
+  static FizzBuzzResponse MakeFizzBuzzResponse(const std::string& data) {
+    FizzBuzzResponse response;
+    response.set_data(data);
+    return response;
+  }
+};
+```
+
+*Notes:*
+
+* There are no unbounded buffers nor any blocking I/O here; it is perfectly fine for a slow client to request a couple of billion responses.
+* It is possible to split any pipeline of Publisher operations into separate functions or even separate RPCs. Backpressure, cancellation and error handling still work transparently.
+* Splitting the plumbing code from logic like this can help make code more readable and testable.
+
+
+#### More
+
+More examples are possible, but this is already getting long. Please ask for more if you want to see what a particular use case would look like. Omitted examples include:
+
+* **Bidirectional streaming:** It is just client and server streaming combined. The handler takes a Publisher and returns a Publisher that can emit more than one element.
+* **Making two upstream calls in parallel:** Can be done with the [`Zip` operator](https://github.com/per-gron/shuriken/blob/master/src/rs/doc/reference.md#zippublisher).
+* **Cancellation:** Most of the time no explicit code related to cancellation is needed. For example, if an RPC has made two upstream requests in parallel and one fails, the other is automatically cancelled. This works with timeouts and streams, making two calls and using just the fastest result ([`Merge()`](https://github.com/per-gron/shuriken/blob/master/src/rs/doc/reference.md#mergepublisher)+[`Take(1)`](https://github.com/per-gron/shuriken/blob/master/src/rs/doc/reference.md#takecount)) etc as well.
+* **Error handling:** With errors, Publishers behave very much like futures/promises do in most languages/implementations. It is basically an async version of `try`/`catch`. Errors are automatically propagated through the chain of operators until they are caught, usually with the `Catch` operator.
+
+
+#### Threading model
+
+The default threading model for a server that uses this library is:
```

Comment:
Why not take an approach ala the ThreadManager used in the synchronous API to allow autosizing the thread count based upon active workload?

Such an approach would eliminate the need for line 313 in this document, and further simplify exposed API.

Reply:
Thanks for the comment! The threading model section is the one I have put the least effort in. I will read and think more about this tomorrow and update the doc and/or reply here.

Reply:
Are you sure that such an approach would not have a negative performance impact? Every test and benchmark I have ever seen has clearly showed that having much more than 1 thread per core causes performance to degrade significantly when the server is under heavy load due to context switches.

I am pretty sure you know more about this stuff than I do so if you say so I will take your word for it but it does surprise me.

Or is the point that doing everything async is best, but if you do have blocking operations, you might as well do it on the same thread pool?